### PR TITLE
Add jinja2 tag

### LINKED ISSUE #7: Jinja2 tag
We use this plugin heavily - and it fecking rocks - would be great if it came with an in built Jinja2 tag 💃

### PR DESCRIPTION
**Pull request by SableWalnut**
Add a Jinja2 tag for videos

Fixes #7.

### CODE REVIEW

File: wagtailvideos/models.py [hunk 1]
```diff
@@ -233,6 +236,23 @@ def get_transcode(self, media_format):
         except Transcode.DoesNotExist:
             return self.do_transcode(media_format)
 
+    def video_tag(self, attrs):
+        kw_attrs = {}
```
Comment (seb-b) [inline wagtailvideos/models.py]: would it be possible for the the regular template tag to use this code too?

File: wagtailvideos/jinja2tags.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+
+def video(video, attrs):
+    if not video:
+        return ''
+
+    if attrs:
+        return video.video_tag(attrs)
+    else:
+        return video
```
Comment (seb-b) [inline wagtailvideos/jinja2tags.py]: will this just return the video object if no attributes are passed?

File: wagtailvideos/jinja2tags.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+from jinja2.ext import Extension
+
+from django import template
+from django.forms.widgets import flatatt
+from django.utils.text import mark_safe
+
+
+def video(video, attrs):
+    if not video:
```
Comment (seb-b) [inline wagtailvideos/jinja2tags.py]: We probably want to error here. An empty string is not very helpful when debugging. The original didn't do this but a type check would also be handy

Comment (timheap) [inline wagtailvideos/jinja2tags.py]: Returning and empty string is fairly traditional, but I agree, it makes it difficult to debug. It will mean doing `{% if foo.video %}{% video foo.video %}{% endif %}` every time you want to use a video though.

File: wagtailvideos/models.py [hunk 4]
```diff
@@ -233,6 +236,23 @@ def get_transcode(self, media_format):
...
+        mime = mimetypes.MimeTypes()
+        sources = ["<source src='{0}' type='{1}'>"
+                   .format(self.url, mime.guess_type(self.url)[0])]
+
+        transcodes = self.transcodes.exclude(processing=True).filter(error_message__exact='')
+        for transcode in transcodes:
+            sources.append("<source src='{0}' type='video/{1}' >".format(transcode.url, transcode.media_format.name))
+        sources.append("<p>Sorry, your browser doesn't support playback for this video</p>")
+        return mark_safe(
+            "<video {0} {1}>{2}</video>".format(attrs, flatatt(kw_attrs), "\n".join(sources)))
```
Comment (timheap) [inline wagtailvideos/models.py]: `attrs` should be passed through `flatatt` as well, so a dict of attrs can be passed in. Also, `flatatt` adds a space before each attr, so you can use it like:
```python
"<video{0}{1}>{2}</video>.format(flatatt(attrs), flatatt(kw_attrs), ...)
```

File: wagtailvideos/models.py (hunk 1 above)
Comment (SableWalnut) [inline wagtailvideos/models.py]: Wouldn't that be in a separate PR?

File: wagtailvideos/models.py (hunk 4 above)
Comment (SableWalnut) [inline wagtailvideos/models.py]: I believe attrs is just a singular string of attrs such as `preload controls autoplay` or something similar, with the attrs not having a value. So it's used like:
```
{{ video(page.video, 'preload controls autoplay') }}
```

File: wagtailvideos/models.py [hunk 5]
```diff
@@ -236,9 +236,7 @@ def get_transcode(self, media_format):
         except Transcode.DoesNotExist:
             return self.do_transcode(media_format)
 
-    def video_tag(self, attrs):
-        kw_attrs = {}
-
+    def video_tag(self, attrs, kw_attrs={}):
```
Comment (timheap) [inline wagtailvideos/models.py]: Having a mutable default argument in a function can lead to surprising bad things: http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments. Better to set it to None, and then do an `if None: kw_attrs={}` in the function body.

File: wagtailvideos/jinja2tags.py [hunk 6]
```diff
@@ -1,22 +1,14 @@
...
 
-def video(video, attrs):
-    if not video:
-        return ''
 
-    if attrs:
-        return video.video_tag(attrs)
-    else:
-        return video
+def video(video, attrs='preload controls'):
```
Comment (timheap) [inline wagtailvideos/jinja2tags.py]: `attrs` should be a dict, so you can do something like:
```jinja2
{{ video(foo.video, bar='baz', quux=foo.some_attr) }}
```
without having to do strange string building. You can have default values for `preload` and `controls` by doing something like:
```python
def video(video, **attrs):
    defaults = {'preload': True, 'controls': True}
    defaults.update(attrs)
    video.video_tag(attrs)
```
This matches how [Wagtail does its `image` tag in Jinja2](https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailimages/jinja2tags.py#L8) as well:

File: wagtailvideos/jinja2tags.py [hunk 7]
```diff
@@ -1,22 +1,14 @@
...
+    if type(video) != Video:
```
Comment (timheap) [inline wagtailvideos/jinja2tags.py]: The pythonic way of doing this is `if isinstance(video, Video):`, which also handles `video` being a subclass of `Video`.

File: wagtailvideos/models.py (hunk 5 above)
Comment (timheap) [inline wagtailvideos/models.py]: With the above comments on the video tag, this would change to become
```python
def video_tag(self, attrs=None):
    if attrs is None:
        attrs = {}
    else:
        attrs = attrs.copy()  # Do not mutate the dict passed in by the caller
    # ...
```
And then use `attrs` instead of `kw_attrs` through the rest of the function.

File: wagtailvideos/models.py [hunk 8]
```diff
@@ -251,7 +249,7 @@ def video_tag(self, attrs):
             sources.append("<source src='{0}' type='video/{1}' >".format(transcode.url, transcode.media_format.name))
         sources.append("<p>Sorry, your browser doesn't support playback for this video</p>")
         return mark_safe(
-            "<video {0} {1}>{2}</video>".format(attrs, flatatt(kw_attrs), "\n".join(sources)))
+            "<video {0}{1}>{2}</video>".format(attrs, flatatt(kw_attrs), "\n".join(sources)))
```
Comment (timheap) [inline wagtailvideos/models.py]: Very minor nit-pick, but this will end up looking like:
```html
<video ...><source ...>
<source ...>
<source ...></video>
```
If you changed the format strings to be like `"<video{attrs}>\n{sources}\n</video>"` then the new lines would be correctly distributed.

File: wagtailvideos/models.py (hunk 4 above)
Comment (timheap) [inline wagtailvideos/models.py]: This is not how the Jinja2 tag for wagtail images works, although there is no need to imitate that exactly. The wagtail images way of doing it does have some benefits though, as I've noted in my review.

**Comment by SableWalnut** [comment]
Should be better now @timheap
